### PR TITLE
Prometheus: Track incremental query request sizes with Faro

### LINKED ISSUE #65214: Prometheus: Incremental querying tracking
**What is the chore?**:
Research a way to implement tracking incremental query savings.

**Is there anything else we need to know?**:
History: Originally Galen hacked a new Grafana header to the incremental query PR, which added a `X-Grafana-Modified-From` header which told us when incremental querying modified a request. However there were a few issues with this approach: it's hard to track these headers in our logs, and that header value doesn't give us the true number of bytes saved, just the number of samples saved, which could scale a lot depending on number of series, or the length of the values.

Looks like @ryantxu opened a PR to add query string params to the api/ds/query calls, but they decided to use headers for backend/proxy reasons.
https://github.com/grafana/grafana/pull/58082

We'd like to make sure that making these urls unique for each incremental query identity (aka panel) won't mess up our current logging, or introduce any other bugs.

### PR DESCRIPTION
**Pull request by leeoniya**
Prototype to sit on top of https://github.com/grafana/grafana/pull/62932
and https://github.com/grafana/grafana/pull/65742

Fixes https://github.com/grafana/grafana/issues/65214

**TL;DR:** Pushes network profiling info from incremental (delta) requests (see #62932) into Faro.

image

**Summary:**
Instead of sending data for each request, we create an accumulator for each panel. Every 5 minutes, or on window unload, we send the status of those accumulators to Faro. 
An example payload:
```
{
  "requestCount": "18", // how many updates has this panel had so far
  "savedBytesTotal": "44879136", // accumulated "savings" in bytes
  "initialRequestSize": "2963337", // initial request size in bytes
  "lastRequestSize": "469643", // the last received request size in bytes
  "panelId": "30", 
  "dashId": "dtngicc4z",
  "expr": "cortex_alertmanager_alerts",
  "interval": "30s"
}
```

**Issues**
Right now we send a request for each panel, which gets kinda network spammy. It shouldn't impact UX as it's all handled asynchronously, but it's not a good use of resources. We need to follow up with the [batch API](https://github.com/grafana/faro-web-sdk/issues/152) when it's available if this instrumentation proves valuable. 

**History**
The reason behind the approach taken here has quite a back-story:

general plan is to:

1. cache the size of the initial full query
2. assume every re-query would have been ~ same size
3. on every incremental query, report the difference between the partial and full query size to Faro

the problem with this is that the backend does not send a `Content-Length` header since we use streaming/chunked JSON dataframe responses. sadly, getting this response length from streamed chunks using web tech gets pretty gnarly and has negative performance implications, such as moving JSON parsing to the main UI thread (synchronous), or cloning the response/stream and other :exploding_head: gymnastics.

as an alternative, i discovered we can get these response length metrics with `PerformanceObserver` much more cheaply, but this is a high-level aggregate/async API that does not provide us with enough details about the underlying Fetch request; all we get is `entry.name`, which is always the same `/api/ds/query` fetch URL. without access to req/resp headers or bodies, we need to somehow tag the requests so we can disambiguate the incremental requests from the normal requests. so i suggested adding some unused query string to the URL so we can then identify it asynchronously, e.g. `/api/ds/query?requestId=Q100`. @ryantxu then pointed me to his failed attempt in https://github.com/grafana/grafana/pull/58082 to add query strings to this endpoint which ended up as headers instead. IMO what we're trying to do here is different than that PR (it't not trying to get the backend to route queries based on querystrings, we just need it for purely frontend perf tracking stuff). my primary concerns with this are:

1. log spam - if we're logging these requests, and each will have a unique key, compressibility may suffer.
2. log queries - if anyone was relying on querying the exact `/api/ds/query` URL instead of a prefix, they would no longer get data. we already have at least one place where we may append a query string, so maybe this is not a concern: 
```js
    let url = '/api/ds/query';
    if (hasExpr) {
      url += '?expression=true';
    }
```
a backend solution would be ideal of course, if it's feasible. we can definitely add headers to the requests that indicate if it's a partial or full request, but the backend would then have to know to record the encoded response size (might be impossible if it's streamed) and log the metrics somewhere. and then the question of getting those metrics into a dashboard.

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 1]
```diff
@@ -62,26 +62,60 @@ function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery>, targ
...
+      this.overlapWindowMs = durationToMilliseconds(duration);
     } else {
       const duration = parseDuration(defaultPrometheusQueryOverlapWindow);
-      this.overlapWindow = durationToMilliseconds(duration);
+      this.overlapWindowMs = durationToMilliseconds(duration);
     }
-  }
 
-  cache = new Map<TargetIdent, TargetCache>();
+    this.perfObeserver = new PerformanceObserver((list) => {
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:83]: IIUC... this intercept all resources after prometheus is initialized 🤔 it does seem like something we may want to move further up the stack for more general reporting (not just prometheus)

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 2]
```diff
@@ -62,26 +62,60 @@ function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery>, targ
...
+
+          if (fetchUrl.includes('/api/ds/query')) {
+            let match = fetchUrl.match(/requestId=([a-z\d]+)/i);
+
+            if (match) {
+              let requestId = match[1];
+
+              if (this.pendingRequestIds.has(requestId)) {
+                // TODO: store full initial request size by targSig so we can diff follow-up partial requests
+                // TODO: log savings between full initial request and incremental request to Faro
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:96]: this seems like somethign that can be done in post processing on loki IIUC

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 1 above)
Comment (leeoniya) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts]: yes, but i wanted to keep it isolated here for now. it't not clear to me if we're gonna get the green light to append requestId to the endpoint url. if that's a blocker, there's no need to spend time to make this generic.

Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts]: Yeah we'll probably wind up feature flagging this, and maybe that can get us the data we want without as much risk?

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 2 above)
Comment (leeoniya) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts]: `requestId`s are not currently unique, they're just sequential from initial dashboard load, e.g. `Q100`, `Q101`. then `Q103` might be initiated by another panel/DS that doesnt need tracking (e.g. Influx), so it will be untagged and omitted from this Prom measuring path.

if we actually had request UIDs, then yeah you can maybe do something in loki. but right now you'll just have a bunch of `Q100`-tagged requests that are unrelated.

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 3]
```diff
@@ -62,26 +64,95 @@ function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery>, targ
...
+                            initialRequestBytes: firstValue.toString(),
+                            subsequentRequestBytes: requestTransferSize.toString(10),
+                          },
+                          'no-interaction',
+                          {
+                            skipDedupe: true,
+                          }
+                        );
+
+                        delete this.pendingRequestIdsToTargSig[requestId];
```
Comment (leeoniya) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:134]: it's better to use a `new Map()` for this and `this.pendingRequestIdsToTargSig.delete(requestId);`

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 4]
```diff
@@ -62,26 +67,112 @@ export function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery
...
  */
-type Milliseconds = number;
 export class QueryCache {
-  private overlapWindow: Milliseconds;
+  private overlapWindowMs;
+  private perfObeserver?: PerformanceObserver;
+  private shouldProfile: boolean;
+
+  pendingRequestIdsToTargSigs = new Map<
+    RequestID,
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:76]: are requestIds unique for each instantiation, or will they repeat at some point?

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 5]
```diff
@@ -62,26 +67,112 @@ export function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery
...
+    if (typeof PerformanceObserver === 'function') {
+      this.perfObeserver = new PerformanceObserver((list: PerformanceObserverEntryList) => {
+        list.getEntries().forEach((entry) => {
+          // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+          const entryTypeCast: PerformanceResourceTiming = entry as PerformanceResourceTiming;
+
+          // Safari support for this is coming in 16.4:
+          // https://caniuse.com/mdn-api_performanceresourcetiming_transfersize
+          // Gating that this exists to prevent runtime errors
+          const isSupported = typeof entryTypeCast?.transferSize === 'number';
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:119]: Good call 👍

File: public/app/plugins/datasource/prometheus/datasource.tsx [hunk 6]
```diff
@@ -455,6 +455,7 @@ export class PrometheusDatasource
       let fullOrPartialRequest: DataQueryRequest<PromQuery>;
       let requestInfo: CacheRequestInfo | undefined = undefined;
       if (this.hasIncrementalQuery) {
+        request.appendRequestId = true;
```
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/datasource.tsx:458]: To cover expressions, I think this part should be copied to `public/app/features/expressions/ExpressionDatasource.ts` as well. Or maybe even having a thing for all datasources. But I see you want to have it only for Prometheus. Is that right?

**Review by itsmylife — COMMENTED** [review]
LGTM 
When it will be ready I will run one last test before approving it.

File: packages/grafana-data/src/types/datasource.ts [hunk 7]
```diff
@@ -494,6 +494,10 @@ export interface DataQueryError {
 export interface DataQueryRequest<TQuery extends DataQuery = DataQuery> {
   requestId: string; // Used to identify results and optionally cancel the request in backendSrv
 
+  // adds requestId querystring param to fetch URL so it can be correlated in an async PerformanceObserver
+  // used for tracking frontend perf metrics, like response sizes, timings, gzip compression, etc.
+  appendRequestId?: boolean;
```
Comment (ryantxu) [inline packages/grafana-data/src/types/datasource.ts:499]: 🤔 i don't like that this is mixed between prometheus specific and has top-level interface changes.

for sure this should not be here -- i think we can add the datasource type to all datasource requests URL (it already has it in headers)

File: public/app/plugins/datasource/prometheus/datasource.tsx (hunk 6 above)
Comment (ryantxu) [reply public/app/plugins/datasource/prometheus/datasource.tsx:458]: lets not copy it anyway... but move it so it is not prometheus specific.

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts (hunk 4 above)
Comment (ryantxu) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:76]: they are unique to each time the PanelQuery runner is called

**Comment by leeoniya** [comment]
> basically giving you the difference in response size between full and partial requests?

yep.

> Temporary to evaluate your new feature works as expected? In that case I think query string is an accepted solution.

well, i think this will stay enabled for a while since the plan is to roll this out gradually, to multiple datasources and eventually make it default. so seeing the full effect would be the ultimate goal. i dont think it will be super-temporary because of this.

**Comment by ryantxu** [comment]
@marefr -- IIUC, the goal here is to understand when this kicks in how much data is "saves" -- the frontend keeps track of the initial request and then subsequent refresh queries -- sending results to faro.  To do the same on the backend, we would either need to log everything and then post-process, or track user sessions; both seem overwhelming.

The url is modified so we can use `PerformanceObserver` on the frontend -- this works great, but requires identifying requests only from the URL, not headers.  I don't see any problem adding url parameters do you?  It is actually kinda nice to see more info in the network tab list view :)

**Comment by marefr** [comment]
Thanks. 

> I don't see any problem adding url parameters do you?

No, not really. It just looked like a lot of work to get this to work for one datasource, hence why I suggested backend.

**Review by gtk-grafana — DISMISSED** [review]
Sends too many events, accumulate savings across panels and batch to Faro every X minutes

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts [hunk 8]
```diff
@@ -64,23 +69,156 @@ export function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery
...
+                }
+              }
+            }
+          }
+        });
+      });
+
+      this.perfObeserver.observe({ type: 'resource', buffered: false });
+
+      setInterval(this.sendPendingTrackingEvents, 60000);
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:186]: Every minute? We could potentially just send events on unload as well

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts
Comment (itsmylife) [inline public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:1]: When I remove `disable-eslint` lines, `eslint` did not complain. So I think they are safe to be removed.

**Review by itsmylife — APPROVED** [review]
LGTM

File: public/app/plugins/datasource/prometheus/querycache/QueryCache.ts
Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/querycache/QueryCache.ts:1]: @itsmylife see the conversation we had with josh on the base branch: https://github.com/grafana/grafana/pull/62932#discussion_r1151964311.

**Review by gtk-grafana — APPROVED** [review]
